git_services/git_services/cli/__init__.py

```diff
@@ -0,0 +1,75 @@
+from pathlib import Path
+import shlex
+from subprocess import Popen, PIPE
+
+
+class GitCommandError(Exception):
+    def __init__(self, returncode, stdout, stderr) -> None:
+        self.returncode = returncode
+        self.stdout = stdout
+        self.stderr = stderr
+
+
+class RepoDirectoryDoesNotExistError(Exception):
+    pass
+
+
+class GitCLI:
```

Comment:
I'm looking at this from a 10000ft view so disregard my question if it's dumb: have you considered GitPython to interface with Git instead of going through the CLI?
https://gitpython.readthedocs.io/en/stable/tutorial.html

I have used it in the past and while sometime is not the easiest library to use, it still works well and quite reliably.

Reply:
GitPython is not actively maintained anymore (just major bugfixes), with the maintainer sometimes just going "Yeah, don't use that feature, it's broken and we won't improve on it" to issues. At least in renku-python we really hope there is a good replacement for it in the future (pygit isn't really an option and dulwich isn't there yet).

Reply:
ah, thanks for the explanation, I was out of the loop.
Is really a big hole in the Python libraries then :/

Reply:
@aledegano yes Ralf mentioned GitPython. The main concern I had with it is that the project is in maintenance mode. It seems the author is working on other things and relying on the community to maintain it. But if there is a bug or something is wrong I am not so sure that this will be addressed that quickly.

Reply:
I am sorry for my late reply. Somehow I completely missed the already present and very detailed explanations from Ralf. I think in my defence they weren't loaded on my page. But it could be that I just missed them.

Reply:
no worries, thank you both for the explanation.